crates/engine/tree/src/tree/mod.rs

```diff
@@ -176,31 +176,115 @@ impl TreeState {
+                // If there are no blocks left then remove the entry for this block
+                if entry.get().is_empty() {
+                    entry.remove();
+                }
+            }
+        }
+
+        Some((executed, children))
+    }
+
+    /// Remove all blocks up to __and including__ the given block number.
+    ///
+    /// If a finalized hash is provided, the only non-canonical blocks which will be removed are
+    /// those which have a fork point at or below the finalized hash.
+    ///
+    /// Canonical blocks below the upper bound will still be removed.
+    ///
+    /// NOTE: This assumes that the `finalized_num` is below or equal to the `upper_bound`
+    pub(crate) fn remove_until(
+        &mut self,
+        upper_bound: BlockNumber,
+        finalized_num: Option<BlockNumber>,
+    ) {
+        debug_assert!(Some(upper_bound) >= finalized_num);
+        // We want to do two things:
+        // * remove canonical blocks that are persisted
+        // * remove forks whose root are below the finalized block
+        // We can do this in 2 steps:
+        // * remove all canonical blocks below the upper bound
+        // * fetch the number of the finalized hash, removing any sidechains that are __below__ the
+        // finalized block
+
+        // TODO: move trie updates here
+        // First, let's walk back the canonical chain and remove canonical blocks lower than the
+        // upper bound
+        let mut current_block = self.current_canonical_head.hash;
+        while let Some(executed) = self.blocks_by_hash.get(&current_block) {
+            current_block = executed.block.parent_hash;
+            if executed.block.number <= upper_bound {
+                self.remove_by_hash(executed.block.hash());
+            }
+        }
+
+        // Now, we have removed canonical blocks (assuming the upper bound is above the finalized
+        // block) and only have sidechains below the finalized block.
+        if let Some(finalized) = finalized_num {
+            // We remove disconnected sidechains in three steps:
+            // * first, remove everything with a block number __below__ the finalized block.
+            // * next, we populate a vec with parents __at__ the finalized block.
+            // * finally, we iterate through the vec, removing children until the vec is empty
+            // (BFS).
+
+            // We _exclude_ the finalized block because we will be dealing with the blocks __at__
+            // the finalized block later.
+            // TODO: remove trie updates whose root are below the finalized block
+            let blocks_to_remove = self
+                .blocks_by_number
+                .range((Bound::Unbounded, Bound::Excluded(finalized)))
+                .flat_map(|(_, blocks)| blocks.iter().map(|b| b.block.hash()))
+                .collect::<Vec<_>>();
+            for hash in blocks_to_remove {
+                self.remove_by_hash(hash);
```

Comment:
should this also remove the child blocks of the removed block?

Reply:
I don't think we need to, we do this anyways in the next loop